biggus/__init__.py

```diff
@@ -561,6 +563,175 @@ def _normalise_keys(self, keys):
+        try:
+            return self.array.ndarray()
+        except AttributeError:
+            return np.array(self.array)
+
+    def masked_array(self):
+        try:
+            return self.array.masked_array()
+        except AttributeError:
+            return np.ma.masked_array(self.array)
+
+
+class NewAxesArray(ArrayContainer):
+    def __init__(self, array, new_axes):
+        """
+        Creates an array which has new axes (i.e. length 1) at the
+        specified locations.
+
+        Parameters
+        ----------
+        array - array like
+            The array upon which to put new axes
+        new_axes - iterable of length array.ndim + 1
+            The number of new axes for each axis.
+            e.g. [2, 1, 0] for a 2d array gain two new axes on the left hand
+            side, one in the middle, and 0 on the right hand side.
+
+        """
+        super(NewAxesArray, self).__init__(array)
+        if self.array.ndim + 1 != len(new_axes):
+            raise ValueError('The new_axes must have length {} but was '
+                             'actually length {}.'.format(self.array.ndim + 1,
+                                                          len(new_axes)))
+        new_axes = np.array(new_axes)
+        dtype_kind = new_axes.dtype.type
+        if (not issubclass(dtype_kind, np.integer) or np.any(new_axes < 0)):
+            raise ValueError('Only positive integer types may be used for '
+                             'new_axes.')
+        self._new_axes = new_axes
+
+    @property
+    def ndim(self):
+        return np.sum(self._new_axes) + self.array.ndim
+
+    @property
+    def shape(self):
+        shape = list(self.array.shape)
+        # Starting from the higher dimensions, insert 1s at the locations
+        # of new axes.
+        for axes, n_new_axes in reversed(list(enumerate(self._new_axes))):
+            for _ in range(n_new_axes):
+                shape.insert(axes, 1)
+        return tuple(shape)
+
+    def _newaxis_keys(self):
+        # Compute the keys needed to produce an array of appropriate newaxis.
+        keys = [slice(None)] * self.array.ndim
+        # Starting from the higher dimensions, insert 1s at the locations
+        # of new axes.
+        for axes, n_new_axes in reversed(list(enumerate(self._new_axes))):
+            for _ in range(n_new_axes):
+                keys.insert(axes, np.newaxis)
+        return tuple(keys)
+
+    def _is_newaxis(self):
+        is_newaxis = [False] * self.array.ndim
+        for axes, n_new_axes in reversed(list(enumerate(self._new_axes))):
+            for _ in range(n_new_axes):
+                is_newaxis.insert(axes, True)
+        return tuple(is_newaxis)
+
+    @classmethod
+    def handle_newaxis(cls, orig_newaxis_func):
+        """
+        A decorator which removes any np.newaxis from ``__getitem__`` keys.
+        If any np.newaxis keys are found, the resulting type from
+        ``__getitem__`` will be a NewAxesArray instance.
```

Comment:
If I were to stand up at a conference and try to explain how you implement a new Array, it is just confusing to not implement the standard python data model (i.e. ``__getitem__``). I fully accept though, if you want to do the job properly, you would still have to explain the decorator.

The obvious choice to avoid both of these is to put it on the metaclass, and always implement the newaxis indexing, unless it is actively disabled.